crates/testing/tests/consensus_task.rs

```diff
@@ -146,6 +147,125 @@ async fn test_consensus_vote() {
+
+    let mut view_2_outputs = vec![
+        exact(ViewChange(ViewNumber::new(2))),
+        exact(QuorumProposalValidated(proposals[1].data.clone())),
+    ];
+    if !force_panic {
+        view_2_outputs.push(exact(QuorumVoteSend(votes[1].clone())));
+    }
+
+    // Use the permuted inputs for view 2 depending on the provided index ordering.
+    let view_2 = TestScriptStage {
+        inputs: view_2_inputs,
+        outputs: view_2_outputs,
+        asserts: vec![is_at_view_number(2)],
+    };
+
+    let consensus_state = ConsensusTaskState::<
+        TestTypes,
+        MemoryImpl,
+        SystemContextHandle<TestTypes, MemoryImpl>,
+    >::create_from(&handle)
+    .await;
+
+    inject_consensus_polls(&consensus_state).await;
+    run_test_script(vec![view_1, view_2], consensus_state).await;
+}
+
+#[cfg(test)]
+#[cfg_attr(
+    async_executor_impl = "tokio",
+    tokio::test(flavor = "multi_thread", worker_threads = 2)
+)]
+#[cfg_attr(async_executor_impl = "async-std", async_std::test)]
+async fn test_consensus_vote_with_permuted_dac() {
+    // These tests verify that a vote is indeed sent no matter when it receives a DACRecv
+    // event. In particular, we want to verify that receiving events in an unexpected (but still
+    // valid) order allows the system to proceed as it normally would.
+    test_vote_with_specific_order(vec![0, 1, 2], false /* force_panic */).await;
+    test_vote_with_specific_order(vec![0, 2, 1], false /* force_panic */).await;
+    test_vote_with_specific_order(vec![1, 0, 2], false /* force_panic */).await;
+    test_vote_with_specific_order(vec![2, 0, 1], false /* force_panic */).await;
+    test_vote_with_specific_order(vec![1, 2, 0], false /* force_panic */).await;
+    test_vote_with_specific_order(vec![2, 1, 0], false /* force_panic */).await;
+}
+
+#[cfg(test)]
+#[cfg_attr(
+    async_executor_impl = "tokio",
+    tokio::test(flavor = "multi_thread", worker_threads = 2)
+)]
+#[cfg_attr(async_executor_impl = "async-std", async_std::test)]
+#[should_panic]
+async fn test_consensus_vote_with_permuted_dac_should_panic() {
+    // These tests verify that a vote is indeed sent no matter when it receives a DACRecv
+    // event. In particular, we want to verify that receiving events in an unexpected (but still
+    // valid) order allows the system to proceed as it normally would. We force the panic here by
+    // intentionally removing the vote send event from the output.
+    test_vote_with_specific_order(vec![0, 1, 2], true /* force_panic */).await;
+    test_vote_with_specific_order(vec![0, 2, 1], true /* force_panic */).await;
+    test_vote_with_specific_order(vec![1, 0, 2], true /* force_panic */).await;
+    test_vote_with_specific_order(vec![2, 0, 1], true /* force_panic */).await;
+    test_vote_with_specific_order(vec![1, 2, 0], true /* force_panic */).await;
+    test_vote_with_specific_order(vec![2, 1, 0], true /* force_panic */).await;
+}
+
```

Comment:
Do these tests actually test anything new? Can we remove the `force_panic` parameter?

Reply:
Sure I'll remove it.